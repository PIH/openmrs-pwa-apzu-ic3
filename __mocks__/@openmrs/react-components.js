import types from 'redux-types';
// mock methods

const visitRest = {

  createVisit: (params) => {

    const visit = params.visit;

    if (visit.patient != 'patient_uuid' || visit.visitType != 'visit_type_uuid' || visit.location != 'location_uuid') {
      throw {
        name: 'Mock Exception',
        message: 'Unable to Submit'
      };
    }
    else {
      return;
    }
  }

};

const encounterRest = {

  createEncounter: (params) => {

    if (params.encounter.encounterType === 'invalid_encounter_type') {
      throw {
        name: 'Mock Exception',
        message: 'Unable to Submit'
      };
    }
    else {
      return;
    }
  }

};

const BASIC_TYPES = [
  'FETCH_SUCCEEDED',
  'FETCH_PENDING',
  'SET_SUCCEEDED',
  'SET_REQUESTED',
  'SET_FAILED',
  'FETCH_FAILED',
  'FETCH_REQUESTED',
  'SUCCEEDED',
  'PENDING',
  'FAILED',
  'REQUESTED'
];

const LOGIN_TYPES = {
  LOGIN: types('login', BASIC_TYPES),
  LOGIN_LOCATIONS: types('login_locations', BASIC_TYPES),
  LOGOUT: types('logout', BASIC_TYPES)
}

const SESSION_TYPES = types('session', BASIC_TYPES)


export {
  visitRest,
  encounterRest,
  LOGIN_TYPES,
  SESSION_TYPES
};
