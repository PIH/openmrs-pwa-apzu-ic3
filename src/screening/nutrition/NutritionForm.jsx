import React from 'react';
import { connect } from 'react-redux';
import { formValueSelector } from 'redux-form';
import {Obs, formValidations, obsRest, formUtil, selectors} from '@openmrs/react-components';
import { Alert, Grid, Row, FormGroup, ControlLabel, Label, Col } from 'react-bootstrap';
import { ENCOUNTER_TYPES, CONCEPTS, MALNUTRITION_LEVEL, FORM_ANSWERS } from "../../constants";
import utils from "../../utils";
import ScreeningForm from "../ScreeningForm";
import { colHeight, leftTextAlign, labelTop } from "../../pwaStyles";

const minValue2 = formValidations.minValue(2);
const minValue20 = formValidations.minValue(20);
const minValue25 = formValidations.minValue(25);
const minValue120 = formValidations.minValue(120);
const maxValue100 = formValidations.maxValue(100);
const maxValue140 = formValidations.maxValue(140);
const maxValue215 = formValidations.maxValue(215);

class NutritionForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      lastHeight: null
    };
  }

  componentDidMount() {
    // measure at each visit for less than 19 y.o., use previous height for >19 years
    if (this.props.patient.age > 19) {
      obsRest.fetchObsByPatient(
        this.props.patient.uuid, CONCEPTS.Height.uuid, 1
      ).then(data => {
        this.setState({
          lastHeight: data.results[0].value
        });
      });
    }
  }

  render() {
    const formContent = (
      <Grid>
        <Row>
          <FormGroup controlId="formWeight">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              Weight
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.Weight.uuid}
                path="weight"
                placeholder="weight in kg"
                validate={this.props.patient.age > 18 ? [minValue25, maxValue140] :  [minValue2, maxValue100]}
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              kg
            </Col>
          </FormGroup>
        </Row>
        <Row>
          <FormGroup controlId="formHeight">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              Height
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.Height.uuid}
                path="height"
                placeholder="height in cm"
                value={ this.state.lastHeight !== null ? this.state.lastHeight : null }
                validate={this.props.patient.age > 18 ? [minValue120, maxValue215] :  [minValue20, maxValue215]}
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              cm
            </Col>
          </FormGroup>
        </Row>

        {( (this.props.patient.age >= 18) && (this.props.patient.age < 50) && (this.props.patient.gender === 'F') ) &&
        <Row>
          <FormGroup controlId="formPregnant">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
            Pregnant
            </Col>
            <Col sm={4}>
              <Obs
                concept={CONCEPTS.Pregnant.uuid}
                conceptAnswers={FORM_ANSWERS.trueFalse}
                path="pregnant"
              />
            </Col>
          </FormGroup>
        </Row>
        }

        { (( this.props.patient.age >= 18 && this.props.patient.gender === 'M')
          || (this.props.patient.gender === 'F' && this.props.patient.age >= 18 && (this.props.pregnant !== null) && (this.props.pregnant === CONCEPTS.False.uuid))
          || (this.props.patient.gender === 'F' && this.props.patient.age > 50))&&
          <Row>
            <FormGroup controlId="formBMI">
              <Col
                componentClass={ControlLabel}
                sm={2}
              >
              BMI
              </Col>

              <Col sm={1}>
                <h3 style={ labelTop }><Label bsStyle={this.props.bmiStyle.alert} style={{visibility: "visible"}}>{ this.props.bmi ? this.props.bmi : "00.00" }</Label>
                </h3>
              </Col>

              <Col
                componentClass={ControlLabel}
                sm={2}
                style={leftTextAlign}
              >
                kg/m<sup>2</sup>
              </Col>
            </FormGroup>
          </Row>
        }

        { ((this.props.patient.age < 18) || ( (this.props.patient.age >= 18) && (this.props.patient.gender === 'F') && (this.props.pregnant !== null) && (this.props.pregnant === CONCEPTS.True.uuid))) &&
        <Row>
          <FormGroup controlId="formMuac">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              MUAC
            </Col>
            <Col sm={2}>
              <Obs
                concept={CONCEPTS.MUAC.uuid}
                path="muac"
                placeholder="muac in cm"
              />
            </Col>
            <Col
              componentClass={ControlLabel}
              sm={2}
              style={leftTextAlign}
            >
              cm
            </Col>
          </FormGroup>
        </Row>
        }
        <Row>
          <FormGroup controlId="formMalnutrition">
            <Col
              componentClass={ControlLabel}
              sm={2}
            >
              <span style={{visibility: this.props.showMalnutrition}}>Malnutrition</span>
            </Col>
            <Col sm={2}>
              <Alert
                bsStyle={this.props.malnutrition ? this.props.malnutrition.alert : "info"}
                style={{visibility: (this.props.malnutrition && this.props.malnutrition.message) ? "visible" : "hidden" }}
              >
                {this.props.malnutrition ? this.props.malnutrition.message : " "}
              </Alert>
            </Col>

          </FormGroup>
        </Row>

        <Row>
          <Col
            md={20}
            sm={20}
            style={colHeight}
          >
            <span><h1>{ '' }</h1></span>
          </Col>
        </Row>
      </Grid>
    );

    return (
      <ScreeningForm
        backLink="/screening/nutrition/queue"
        encounterType={ENCOUNTER_TYPES.NutritionEncounterType}
        formContent={formContent}
        formId="nutrition-form"
        formInstanceId="nutrition-form"
        defaultValues={ this.state.lastHeight !== null ? [{
          type: "obs",
          path: "height",
          concept: CONCEPTS.Height.uuid,
          value: this.state.lastHeight
        }] : null
        }
        title="Nutrition"
        toastMessage="Nutrition Saved"
      />
    );
  }

}

const selector = formValueSelector('nutrition-form');

export default connect(state => {
  const patient = selectors.getSelectedPatientFromStore(state);
  const weight = selector(state, formUtil.obsFieldName('weight', CONCEPTS.Weight.uuid));
  const height = selector(state, formUtil.obsFieldName('height', CONCEPTS.Height.uuid));
  const muac = selector(state, formUtil.obsFieldName('muac', CONCEPTS.MUAC.uuid));
  const pregnant = selector(state, formUtil.obsFieldName('pregnant', CONCEPTS.Pregnant.uuid));
  let bmi = null;
  let bmiStyle = MALNUTRITION_LEVEL.none;
  if ( typeof weight !== 'undefined' && typeof height !== 'undefined') {
    bmi = utils.calculateBMI(weight, height);
    bmiStyle = utils.calculateBMIAlert(bmi);
  }
  const malnutrition = utils.calculateMalnutritionLevel(bmi, muac, patient ? patient.age : null, pregnant);
  let showMalnutrition = "hidden";
  if (malnutrition &&
    (malnutrition === MALNUTRITION_LEVEL.moderate || malnutrition === MALNUTRITION_LEVEL.severe)) {
    showMalnutrition = "visible";
  }
  return {
    weight,
    height,
    pregnant,
    bmi,
    bmiStyle,
    malnutrition,
    showMalnutrition,
    patient,
  };
})(NutritionForm);

