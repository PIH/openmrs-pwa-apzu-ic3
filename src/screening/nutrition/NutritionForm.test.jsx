import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { OpenMRSForm } from '@openmrs/react-components';
import NutritionForm from './NutritionForm';

let props, store;
let mountedComponent;

const mockStore = configureMockStore();

const nutritionForm = () => {
  if (!mountedComponent) {
    mountedComponent = mount(
      <Provider store={store}>
        <NutritionForm {...props} />
      </Provider>);
  }
  return mountedComponent;
};

describe('Component: nutritionForm', () => {
  beforeEach(() => {
    props = {};
    store = mockStore(
      {
        selected: {
          patient: {
            patient: {
              uuid: "abc"
            },
            visit: {
              uuid: "def"
            }
          },
        }
      });
    mountedComponent = undefined;
  });

  it('renders properly', () => {
    expect(toJson(nutritionForm())).toMatchSnapshot();
    expect(nutritionForm().find(OpenMRSForm).length).toBe(1);
  });

});
