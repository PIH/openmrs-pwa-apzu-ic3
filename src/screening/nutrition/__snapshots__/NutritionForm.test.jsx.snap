// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: nutritionForm renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(NutritionForm)>
    <NutritionForm
      dispatch={[Function]}
      patient={
        Object {
          "uuid": "abc",
        }
      }
      visit={
        Object {
          "uuid": undefined,
        }
      }
    >
      <Connect(ReduxForm)
        encounterType={
          Object {
            "uuid": "4fb47712-34a6-40d2-8ed3-e153abbd25b7",
          }
        }
        formSubmittedActionCreator={[Function]}
        patient={
          Object {
            "uuid": "abc",
          }
        }
        visit={
          Object {
            "uuid": undefined,
          }
        }
      >
        <ReduxForm
          dispatch={[Function]}
          encounterType={
            Object {
              "uuid": "4fb47712-34a6-40d2-8ed3-e153abbd25b7",
            }
          }
          formSubmittedActionCreator={[Function]}
          patient={
            Object {
              "uuid": "abc",
            }
          }
          visit={
            Object {
              "uuid": undefined,
            }
          }
        >
          <Connect(Form(OpenMRSForm))
            destroyOnUnmount={true}
            dispatch={[Function]}
            enableReinitialize={false}
            encounterType={
              Object {
                "uuid": "4fb47712-34a6-40d2-8ed3-e153abbd25b7",
              }
            }
            forceUnregisterOnUnmount={false}
            form="openmrs-form"
            formSubmittedActionCreator={[Function]}
            getFormState={[Function]}
            keepDirtyOnReinitialize={false}
            patient={
              Object {
                "uuid": "abc",
              }
            }
            persistentSubmitErrors={false}
            pure={true}
            shouldAsyncValidate={[Function]}
            shouldError={[Function]}
            shouldValidate={[Function]}
            shouldWarn={[Function]}
            touchOnBlur={true}
            touchOnChange={false}
            updateUnregisteredFields={false}
            visit={
              Object {
                "uuid": undefined,
              }
            }
          >
            <Form(OpenMRSForm)
              anyTouched={false}
              array={
                Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "removeAll": [Function],
                  "shift": [Function],
                  "splice": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                }
              }
              arrayInsert={[Function]}
              arrayMove={[Function]}
              arrayPop={[Function]}
              arrayPush={[Function]}
              arrayRemove={[Function]}
              arrayRemoveAll={[Function]}
              arrayShift={[Function]}
              arraySplice={[Function]}
              arraySwap={[Function]}
              arrayUnshift={[Function]}
              asyncValidating={false}
              autofill={[Function]}
              blur={[Function]}
              change={[Function]}
              clearAsyncError={[Function]}
              clearFields={[Function]}
              clearSubmit={[Function]}
              clearSubmitErrors={[Function]}
              destroy={[Function]}
              destroyOnUnmount={true}
              dirty={false}
              dispatch={[Function]}
              enableReinitialize={false}
              encounterType={
                Object {
                  "uuid": "4fb47712-34a6-40d2-8ed3-e153abbd25b7",
                }
              }
              focus={[Function]}
              forceUnregisterOnUnmount={false}
              form="openmrs-form"
              formSubmittedActionCreator={[Function]}
              getFormState={[Function]}
              initialize={[Function]}
              initialized={false}
              invalid={false}
              keepDirtyOnReinitialize={false}
              patient={
                Object {
                  "uuid": "abc",
                }
              }
              persistentSubmitErrors={false}
              pristine={true}
              pure={true}
              registerField={[Function]}
              reset={[Function]}
              resetSection={[Function]}
              setSubmitFailed={[Function]}
              setSubmitSucceeded={[Function]}
              shouldAsyncValidate={[Function]}
              shouldError={[Function]}
              shouldValidate={[Function]}
              shouldWarn={[Function]}
              startAsyncValidation={[Function]}
              startSubmit={[Function]}
              stopAsyncValidation={[Function]}
              stopSubmit={[Function]}
              submit={[Function]}
              submitFailed={false}
              submitSucceeded={false}
              submitting={false}
              syncErrors={Object {}}
              syncWarnings={Object {}}
              touch={[Function]}
              touchOnBlur={true}
              touchOnChange={false}
              unregisterField={[Function]}
              untouch={[Function]}
              updateSyncErrors={[Function]}
              updateSyncWarnings={[Function]}
              updateUnregisteredFields={false}
              valid={true}
              validExceptSubmit={true}
              values={Object {}}
              visit={
                Object {
                  "uuid": undefined,
                }
              }
            >
              <OpenMRSForm
                anyTouched={false}
                array={
                  Object {
                    "insert": [Function],
                    "move": [Function],
                    "pop": [Function],
                    "push": [Function],
                    "remove": [Function],
                    "removeAll": [Function],
                    "shift": [Function],
                    "splice": [Function],
                    "swap": [Function],
                    "unshift": [Function],
                  }
                }
                asyncValidate={[Function]}
                asyncValidating={false}
                autofill={[Function]}
                blur={[Function]}
                change={[Function]}
                clearAsyncError={[Function]}
                clearFields={[Function]}
                clearSubmit={[Function]}
                clearSubmitErrors={[Function]}
                destroy={[Function]}
                dirty={false}
                dispatch={[Function]}
                encounterType={
                  Object {
                    "uuid": "4fb47712-34a6-40d2-8ed3-e153abbd25b7",
                  }
                }
                form="openmrs-form"
                formSubmittedActionCreator={[Function]}
                handleSubmit={[Function]}
                initialize={[Function]}
                initialized={false}
                invalid={false}
                patient={
                  Object {
                    "uuid": "abc",
                  }
                }
                pristine={true}
                pure={true}
                reset={[Function]}
                resetSection={[Function]}
                submit={[Function]}
                submitFailed={false}
                submitSucceeded={false}
                submitting={false}
                touch={[Function]}
                untouch={[Function]}
                valid={true}
                visit={
                  Object {
                    "uuid": undefined,
                  }
                }
              >
                <Form
                  bsClass="form"
                  componentClass="form"
                  horizontal={true}
                  inline={false}
                  onSubmit={[Function]}
                >
                  <form
                    className="form-horizontal"
                    onSubmit={[Function]}
                  >
                    <Grid
                      bsClass="container"
                      componentClass="div"
                      fluid={false}
                    >
                      <div
                        className="container"
                      >
                        <Row
                          bsClass="row"
                          componentClass="div"
                        >
                          <div
                            className="row"
                          >
                            Add Nutrituion Form
                          </div>
                        </Row>
                        <Row
                          bsClass="row"
                          componentClass="div"
                        >
                          <div
                            className="row"
                          >
                            <Submit>
                              <Button
                                active={false}
                                block={false}
                                bsClass="btn"
                                bsStyle="success"
                                disabled={false}
                                type="submit"
                              >
                                <button
                                  className="btn btn-success"
                                  disabled={false}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Button>
                            </Submit>
                          </div>
                        </Row>
                      </div>
                    </Grid>
                  </form>
                </Form>
              </OpenMRSForm>
            </Form(OpenMRSForm)>
          </Connect(Form(OpenMRSForm))>
        </ReduxForm>
      </Connect(ReduxForm)>
    </NutritionForm>
  </Connect(NutritionForm)>
</Provider>
`;
